<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Overview &mdash; MHC-Matchmaker  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=7dc6ce7e" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Worked out example" href="example.html" />
    <link rel="prev" title="MHC-Matchmaker documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            MHC-Matchmaker
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-data">Input Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alignment">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#grouping-the-alleles">Grouping the alleles</a></li>
<li class="toctree-l2"><a class="reference internal" href="#counting-initial-mismatches">Counting initial mismatches</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#single-position-mismatches">Single position mismatches</a></li>
<li class="toctree-l3"><a class="reference internal" href="#full-sequence-mismatches">Full sequence mismatches</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mismatch-significance">Mismatch significance</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#solvent-accessibility-scores">Solvent accessibility scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rankings">Rankings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#identifying-known-eplets">Identifying known eplets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example.html">Worked out example</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHC-Matchmaker</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Overview</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Link to this heading"></a></h1>
<p>This project evaluates and ranks Major Histocompatibility Complex (MHC) compatibility between donors and recipients across multiple species, including humans, pigs, and various macaque species (Macaca mulatta and Macaca fascicularis).</p>
<p><img alt="image" src="_images/nhp.png" /> <img alt="image" src="_images/sla.png" /></p>
<!-- <-- This is done by:

1. Aligning the input sequences
2. Calculating mismatches
3. Filtering the mismatches to relevant mismatches by using solvent accessibility scores
4. Identifying known HLA eplets among the mismatches --> 
<section id="input-data">
<h2>Input Data<a class="headerlink" href="#input-data" title="Link to this heading"></a></h2>
<p>The input is a collection of donors and recipients each characterized by a set of MHC alleles. <br />
These alleles are sourced from the Immuno Polymorphism Database <a class="reference external" href="https://www.ebi.ac.uk/ipd">(IPD)</a>.<br />
Currently the project supports alleles from the following species:</p>
<ul class="simple">
<li><p>Humans (HLA - Human Leukocyte Antigen)</p></li>
<li><p>Pigs (SLA - Swine Leukocyte Antigen)</p></li>
<li><p>Non-Human Primates:</p>
<ul>
<li><p>Rhesus macaques (Mamu - Macaca mulatta)</p></li>
<li><p>Crab-eating macaques (Mafa - Macaca fascicularis)</p></li>
</ul>
</li>
</ul>
<p>Throughout this project, the donor and recipient alleles are kept classified in their respective classes:</p>
<ul class="simple">
<li><p>Class I</p></li>
<li><p>Class II (DQA, DQB, DRA, DRB)</p></li>
</ul>
<p>Only alleles present in the <a class="reference external" href="https://www.ebi.ac.uk/ipd">IPD-IMGT/HLA database</a> are considered to be valid input.
However, if the allele is not present in the database but it:</p>
<ul class="simple">
<li><p>was used as previous nomenclature</p></li>
<li><p>can be extended/shortened to a valid allele. e.g. “Mamu-A11:01:01” can be shortened to “Mamu-A11:01”.
It will be corrected/replaced with a valid allele (<a class="reference internal" href="example.html#input-data"><span class="std std-ref">see example</span></a>).</p></li>
</ul>
<p>The status of the allele is also checked to see if the allele hasn’t been <a class="reference external" href="https://www.ebi.ac.uk/ipd/mhc/version/v31300/#abandoned">abandoned</a>.</p>
</section>
<section id="alignment">
<h2>Alignment<a class="headerlink" href="#alignment" title="Link to this heading"></a></h2>
<p>All alleles belonging to the same class are aligned to a reference consensus sequence.<br />
(TO DO: cite the origin of consensus sequences)</p>
<!-- TO DO: ADD REFERENCE\ -->
<p>Alleles of the same class are aligned to a reference consensus sequence using MAFFT <a class="reference internal" href="#1"><span class="xref myst">[1]</span></a> with the following options:</p>
<ul class="simple">
<li><p>–add: adds unaligned full length sequences to an existing alignment.</p></li>
<li><p>–keeplength: maintains the length of the existing alignment.</p></li>
</ul>
<p>For more details, see the MAFFT <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/addsequences.html">documentation</a>.</p>
</section>
<section id="grouping-the-alleles">
<h2>Grouping the alleles<a class="headerlink" href="#grouping-the-alleles" title="Link to this heading"></a></h2>
<p>Every donor and recipient is characterized by a list of alleles per class. To be able to compare the donor and recipient, the alleles are grouped by position.<br />
This is done as follows:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Position</p></th>
<th class="head text-center"><p>1</p></th>
<th class="head text-center"><p>2</p></th>
<th class="head text-center"><p>3</p></th>
<th class="head text-center"><p>4</p></th>
<th class="head text-center"><p>5</p></th>
<th class="head text-center"><p>6</p></th>
<th class="head text-center"><p>7</p></th>
<th class="head text-center"><p>8</p></th>
<th class="head text-center"><p>9</p></th>
<th class="head text-center"><p>10</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Allele 1</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>K</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>V</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
<tr class="row-odd"><td><p>Allele 2</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>R</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>I</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
<tr class="row-even"><td><p>Grouped allele</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>KR</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>VI</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
</tbody>
</table>
</section>
<section id="counting-initial-mismatches">
<h2>Counting initial mismatches<a class="headerlink" href="#counting-initial-mismatches" title="Link to this heading"></a></h2>
<p>Before diving into counting mismatches for entire grouped donor and recipient alleles, let’s see how mismatches are counted at a single position of those grouped sequences.</p>
<section id="single-position-mismatches">
<h3>Single position mismatches<a class="headerlink" href="#single-position-mismatches" title="Link to this heading"></a></h3>
<p>Taking the donor and recipient sequences with  amino acids at position i below:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head text-center"><p>Allele 1</p></th>
<th class="head text-center"><p>Allele 2</p></th>
<th class="head text-center"><p>Allele 3</p></th>
<th class="head text-center"><p>Allele 4</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Amino acids of donor sequences at position i</p></td>
<td class="text-center"><p>B</p></td>
<td class="text-center"><p>B</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>B</p></td>
</tr>
<tr class="row-odd"><td><p>Amino acids of recipient sequences at position i</p></td>
<td class="text-center"><p>R</p></td>
<td class="text-center"><p>K</p></td>
<td class="text-center"><p>R</p></td>
<td class="text-center"><p>A</p></td>
</tr>
</tbody>
</table>
<p>The donor mismatches are defined as the amino acids that are present in the donor but absent in the  recipient.
For the example above the donor mismatch is “B”.</p>
<p>The recipient mismatches are defined as the amino acids that are present in the recipient but absent in the donor.
For the example above the recipient mismatches are “R” and “K”.</p>
<p>The figure below illustrates the concept of donor/recipient mismatches through set differences and intersections.</p>
<figure class="align-center">
<img alt="singleMismatch" src="_images/singleMismatch.png" />
</figure>
</section>
<section id="full-sequence-mismatches">
<h3>Full sequence mismatches<a class="headerlink" href="#full-sequence-mismatches" title="Link to this heading"></a></h3>
<p>To count the donor/recipient mismatches of the entire grouped sequences, the single position mismatches are calculated for each position in the sequence.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Position</p></th>
<th class="head text-center"><p>1</p></th>
<th class="head text-center"><p>2</p></th>
<th class="head text-center"><p>3</p></th>
<th class="head text-center"><p>4</p></th>
<th class="head text-center"><p>5</p></th>
<th class="head text-center"><p>6</p></th>
<th class="head text-center"><p>7</p></th>
<th class="head text-center"><p>8</p></th>
<th class="head text-center"><p>9</p></th>
<th class="head text-center"><p>10</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Grouped Recipient seq.</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>AKR</p></td>
<td class="text-center"><p>IE</p></td>
<td class="text-center"><p>YV</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
<tr class="row-odd"><td><p>Grouped Donor seq.</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>AB</p></td>
<td class="text-center"><p>E</p></td>
<td class="text-center"><p>VI</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>N</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
<tr class="row-even"><td><p>Donor mismatches</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>B</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>I</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>N</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
</tr>
<tr class="row-odd"><td><p>Recipient mismatches</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>KR</p></td>
<td class="text-center"><p>I</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="mismatch-significance">
<h3>Mismatch significance<a class="headerlink" href="#mismatch-significance" title="Link to this heading"></a></h3>
<p>The mismatch significance is an additional ratio that can be used to determine the importance of a mismatch.
It is calculated by:</p>
<ul class="simple">
<li><p>Donor mismatch significance: Number of donor sequences that have a mismatch/total number of donor alleles</p></li>
<li><p>Recipient mismatch significance: Number of recipient sequences that have a mismatch/total number of recipient alleles</p></li>
</ul>
<p>Going back to the example shown above, let’s calculate the donor mismatch significance score at position 2:<br />
The donor has a total of 4 alleles, with 3 alleles having the amino acid B and a single allele having the amino acid A.
This gives us a ratio of 0.75.</p>
</section>
</section>
<section id="solvent-accessibility-scores">
<h2>Solvent accessibility scores<a class="headerlink" href="#solvent-accessibility-scores" title="Link to this heading"></a></h2>
<p>Not all amino acid mismatches are equally relevant as some positions are buried in the 3D protein structure. To filter relevant mismatches, solvent accessibility scores are used to determine if an amino acid is buried or exposed in the protein structure.<br />
To calculate the SAS score the <a class="reference external" href="https://github.com/jas-preet/SPOT-1D-Single">SPOT-1D</a> model was used to predict relative surface accessibility (RSA) scores <a class="reference internal" href="#2"><span class="xref myst">[2]</span></a>.</p>
<p>An example of RSA scores for a single allele are shown below:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Position</p></th>
<th class="head text-center"><p>1</p></th>
<th class="head text-center"><p>2</p></th>
<th class="head text-center"><p>3</p></th>
<th class="head text-center"><p>4</p></th>
<th class="head text-center"><p>5</p></th>
<th class="head text-center"><p>6</p></th>
<th class="head text-center"><p>7</p></th>
<th class="head text-center"><p>8</p></th>
<th class="head text-center"><p>9</p></th>
<th class="head text-center"><p>10</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Allele</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>K</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>V</p></td>
<td class="text-center"><p>F</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>Y</p></td>
<td class="text-center"><p>D</p></td>
<td class="text-center"><p>S</p></td>
<td class="text-center"><p>R</p></td>
</tr>
<tr class="row-odd"><td><p>RSA score</p></td>
<td class="text-center"><p>0.85</p></td>
<td class="text-center"><p>0.48</p></td>
<td class="text-center"><p>0.49</p></td>
<td class="text-center"><p>0.22</p></td>
<td class="text-center"><p>0.54</p></td>
<td class="text-center"><p>0.43</p></td>
<td class="text-center"><p>0.43</p></td>
<td class="text-center"><p>0.21</p></td>
<td class="text-center"><p>0.31</p></td>
<td class="text-center"><p>0.12</p></td>
</tr>
</tbody>
</table>
<p>To calculate the RSA scores for an entire donor/recipient, the individual allele scores are averaged over all the donor/recipient alleles at each position.</p>
<p>An example of calculating the grouped RSA scores for a donor/recipient are shown below:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Position</p></th>
<th class="head text-center"><p>1</p></th>
<th class="head text-center"><p>2</p></th>
<th class="head text-center"><p>3</p></th>
<th class="head text-center"><p>4</p></th>
<th class="head text-center"><p>5</p></th>
<th class="head text-center"><p>6</p></th>
<th class="head text-center"><p>7</p></th>
<th class="head text-center"><p>8</p></th>
<th class="head text-center"><p>9</p></th>
<th class="head text-center"><p>10</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Seq. 1 RSA scores</p></td>
<td class="text-center"><p>0.72</p></td>
<td class="text-center"><p>0.95</p></td>
<td class="text-center"><p>0.38</p></td>
<td class="text-center"><p>-</p></td>
<td class="text-center"><p>0.29</p></td>
<td class="text-center"><p>0.84</p></td>
<td class="text-center"><p>0.47</p></td>
<td class="text-center"><p>0.13</p></td>
<td class="text-center"><p>0.56</p></td>
<td class="text-center"><p>0.90</p></td>
</tr>
<tr class="row-odd"><td><p>Seq. 2 RSA scores</p></td>
<td class="text-center"><p>0.83</p></td>
<td class="text-center"><p>0.27</p></td>
<td class="text-center"><p>0.61</p></td>
<td class="text-center"><p>0.15</p></td>
<td class="text-center"><p>0.92</p></td>
<td class="text-center"><p>0.46</p></td>
<td class="text-center"><p>-</p></td>
<td class="text-center"><p>0.38</p></td>
<td class="text-center"><p>0.05</p></td>
<td class="text-center"><p>0.70</p></td>
</tr>
<tr class="row-even"><td><p>Grouped RSA score</p></td>
<td class="text-center"><p>0.775</p></td>
<td class="text-center"><p>0.61</p></td>
<td class="text-center"><p>0.495</p></td>
<td class="text-center"><p>0.15</p></td>
<td class="text-center"><p>0.605</p></td>
<td class="text-center"><p>0.65</p></td>
<td class="text-center"><p>0.47</p></td>
<td class="text-center"><p>0.255</p></td>
<td class="text-center"><p>0.305</p></td>
<td class="text-center"><p>0.8</p></td>
</tr>
</tbody>
</table>
<!-- Possible improvement -> Go check which sequences are responsible for the mismatch and check those sequences for their solvent accessibility at the mismatch position. -->
<p>To filter the mismatches based on the solvent accessibility scores, a threshold can be set, scores below the threshold are filtered out.</p>
</section>
<section id="rankings">
<h2>Rankings<a class="headerlink" href="#rankings" title="Link to this heading"></a></h2>
<p>With the mismatches and their RSA scores calculated, donors can be ranked based on:</p>
<ul class="simple">
<li><p>The number of all initial donor or recipient mismatches</p></li>
<li><p>The number of RSA filtered donor or recipient mismatches</p></li>
</ul>
</section>
<section id="identifying-known-eplets">
<h2>Identifying known eplets<a class="headerlink" href="#identifying-known-eplets" title="Link to this heading"></a></h2>
<p>This section aims to identify known HLA eplets that are present in donor or recipient mismatches.
The known eplets are taken from the <a class="reference external" href="https://www.epregistry.com.br">HLA Eplet Registry</a> and are filtered to only include the confirmed eplets.</p>
<p>To identify known eplets, the following is done (handle the case for donor, but the same can be done for recipient):<br />
For each donor mismatch, go over the alleles that are responsible for the mismatch. Go over the eplets and identify known eplets that have the mismatch position as part of the eplet. Finally, check which eplets are present in the donor alleles responsible for the mismatch.</p>
<p><strong><em>NOTE:</em></strong> The mismatches used to identify the eplets are the unfiltered/initial mismatches.</p>
<p>Take the following example, where there is a donor mismatch at position 44:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Position</p></th>
<th class="head text-center"><p>…</p></th>
<th class="head text-center"><p>42</p></th>
<th class="head text-center"><p>43</p></th>
<th class="head text-center"><p>44</p></th>
<th class="head text-center"><p>45</p></th>
<th class="head text-center"><p>46</p></th>
<th class="head text-center"><p>…</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Donor seq.</p></td>
<td class="text-center"><p>…</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>K</p></td>
<td class="text-center"><p>R</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>…</p></td>
</tr>
<tr class="row-odd"><td><p>Recipient seq.</p></td>
<td class="text-center"><p>…</p></td>
<td class="text-center"><p>L</p></td>
<td class="text-center"><p>K</p></td>
<td class="text-center"><p>M</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p>…</p></td>
</tr>
<tr class="row-even"><td><p>Donor Mismatch</p></td>
<td class="text-center"><p>…</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>R</p></td>
<td class="text-center"><p>A</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>…</p></td>
</tr>
</tbody>
</table>
<p>Eplet 44RMA is a confirmed eplet that consists of the amino acids R, M and A at position 44, 45 and 46 respectively. <br />
The donor at position 44 causes a mismatch, the known eplet (44RMA) is then identified at that position.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading"></a></h2>
<p><a id="1">[1]</a>
Katoh, K., Rozewicki, J., &amp; Yamada, K. D. (2019). MAFFT online service: Multiple sequence alignment, interactive sequence choice and visualization. Briefings in Bioinformatics, 20(4), 1160–1166. https://doi.org/10.1093/bib/bbx108</p>
<p><a id="2">[2]</a>
Singh, J., Litfin, T., Paliwal, K., Singh, J., Hanumanthappa, A. K., &amp; Zhou, Y. (2021). SPOT-1D-Single: Improving the single-sequence-based prediction of protein secondary structure, backbone angles, solvent accessibility and half-sphere exposures using a large training set and ensembled deep learning. Bioinformatics, 37(20), 3464–3472. https://doi.org/10.1093/bioinformatics/btab316</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="MHC-Matchmaker documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="example.html" class="btn btn-neutral float-right" title="Worked out example" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>